File: Java/binary_search/search_Two_Dimensional_Matrix.java
public class search_Two_Dimensional_Matrix {


    public static void main(String[] args){
        int[][]matrix = {{1,2,4,8},{10,11,12,13},{14,20,30,40}};
        int target = 8;
    }

    /**
     * Leetcode # 74 - medium
     * Search a 2D Matrix
     * @param matrix 2D num array
     * @param target int value
     * @return true or false
     */
    public static boolean searchMatrix(int[][] matrix, int target){
        int top = 0;
        int  bottom = matrix.length - 1;
        int col_right = matrix[0].length - 1;
        int col_left = 0;
        int col_mid;
        int middleRow = 0;
        while(top <= bottom){
            middleRow = (top  + bottom)/2;
            if(target >= matrix[middleRow][0] && target <= matrix[middleRow][col_right]){
                break;
            }
            else if(matrix[middleRow][0] < target){
                top = middleRow + 1;
            } else {
                bottom = middleRow - 1;
            }
        }
        if(top > bottom){
            return false; // No row found
        }
        while(col_left <= col_right){
            col_mid = (col_right + col_left)/2;
            if(target < matrix[middleRow][col_mid]){
                col_right = col_mid - 1;
            } else  if(target > matrix[middleRow][col_mid]){
                col_left = col_mid + 1;
            } else {
                return true;
            }
        }
        return false;
    }
}

File: Python/binary_search/search_two_dimensional_matrix.py

# LeetCode (74) Search a 2D matrix
# Medium
def searchMatrix(matrix: list[list[int]], target: int) -> bool:
    top_row = 0
    bottom_row = len(matrix) - 1
    middle_row = 0
    col_right = len(matrix[0]) - 1
    col_left = 0
    col_mid = 0
    while top_row <= bottom_row:
        middle_row = (bottom_row + top_row)//2
        if matrix[middle_row][0] <= target <= matrix[middle_row][col_right]:
            break
        elif target > matrix[middle_row][0]:
            top_row = middle_row + 1
        else :
            bottom_row = middle_row - 1
    if top_row > bottom_row:
        return False
    while col_left <= col_right:
        col_mid  = (col_right + col_left) // 2
        if target < matrix[middle_row][col_mid]:
            col_right = col_mid  - 1
        elif target > matrix[middle_row][col_mid]:
            col_left = col_mid + 1
        else:
            return True
    return False
